tests/gocase/unit/type/strings/strings_test.go

```diff
@@ -0,0 +1,649 @@
+		require.EqualValues(t, 4, rdb.SetRange(ctx, "mykey", 1, "3").Val())
+		require.Equal(t, "1334", rdb.Get(ctx, "mykey").Val())
+
+		require.NoError(t, rdb.Set(ctx, "mykey", 1234, 0).Err())
+		require.EqualValues(t, 6, rdb.SetRange(ctx, "mykey", 5, "2").Val())
+		require.Equal(t, "1234\0002", rdb.Get(ctx, "mykey").Val())
+	})
+
+	t.Run("SETRANGE against key with wrong type", func(t *testing.T) {
+		require.NoError(t, rdb.Del(ctx, "mykey").Err())
+		require.NoError(t, rdb.LPush(ctx, "mykey", "foo").Err())
+		r := rdb.SetRange(ctx, "mykey", 0, "bar")
+		require.ErrorContains(t, r.Err(), "WRONGTYPE")
+	})
+
+	t.Run("GETRANGE against non-existing key", func(t *testing.T) {
+		require.NoError(t, rdb.Del(ctx, "mykey").Err())
+		require.EqualValues(t, "", rdb.GetRange(ctx, "mykey", 0, -1).Val())
+	})
+
+	t.Run("GETRANGE against string value", func(t *testing.T) {
+		require.NoError(t, rdb.Set(ctx, "mykey", "Hello World", 0).Err())
+		require.EqualValues(t, "Hell", rdb.GetRange(ctx, "mykey", 0, 3).Val())
+		require.EqualValues(t, "ll", rdb.GetRange(ctx, "mykey", 2, 3).Val())
+		require.EqualValues(t, "Hello World", rdb.GetRange(ctx, "mykey", 0, -1).Val())
+		require.EqualValues(t, "orld", rdb.GetRange(ctx, "mykey", -4, -1).Val())
+		require.EqualValues(t, "", rdb.GetRange(ctx, "mykey", 5, 3).Val())
+		require.EqualValues(t, " World", rdb.GetRange(ctx, "mykey", 5, 5000).Val())
+		require.EqualValues(t, "Hello World", rdb.GetRange(ctx, "mykey", -5000, 10000).Val())
+	})
+
+	t.Run("GETRANGE against integer-encoded value", func(t *testing.T) {
+		require.NoError(t, rdb.Set(ctx, "mykey", 1234, 0).Err())
+		require.EqualValues(t, "123", rdb.GetRange(ctx, "mykey", 0, 2).Val())
+		require.EqualValues(t, "1234", rdb.GetRange(ctx, "mykey", 0, -1).Val())
+		require.EqualValues(t, "234", rdb.GetRange(ctx, "mykey", -3, -1).Val())
+		require.EqualValues(t, "", rdb.GetRange(ctx, "mykey", 5, 3).Val())
+		require.EqualValues(t, "4", rdb.GetRange(ctx, "mykey", 3, 5000).Val())
+		require.EqualValues(t, "1234", rdb.GetRange(ctx, "mykey", -5000, 10000).Val())
+	})
+
+	t.Run("Extended SET can detect syntax errors", func(t *testing.T) {
+		require.NoError(t, c.WriteArgs("SET", "foo", "bar", "non-existing-option"))
+		r, err := c.ReadLine()
+		require.NoError(t, err)
+		require.Contains(t, r, "syntax error")
+	})
+
+	t.Run("Extended SET NX option", func(t *testing.T) {
+		require.NoError(t, rdb.Del(ctx, "foo").Err())
+
+		require.NoError(t, c.WriteArgs("SET", "foo", "1", "nx"))
+		r, err := c.ReadLine()
+		require.NoError(t, err)
+		require.Equal(t, "+OK", r)
+
+		require.NoError(t, c.WriteArgs("SET", "foo", "2", "nx"))
+		r, err = c.ReadLine()
+		require.NoError(t, err)
+		require.Equal(t, "$-1", r)
+
+		require.Equal(t, "1", rdb.Get(ctx, "foo").Val())
+	})
+
+	t.Run("Extended SET XX option", func(t *testing.T) {
+		require.NoError(t, rdb.Del(ctx, "foo").Err())
+
+		require.NoError(t, c.WriteArgs("SET", "foo", "1", "xx"))
```

Comment:
Please try to use `redis.Client`'s method instead of sending commands with raw client.

ditto all others

Reply:
`(redis.Client).Set` does not support `Extended SET XX option`, it only take `context`, `key`, `value` and `expiration` as arguements, so i use raw client to work around.

Reply:
and (redis.Client).Set` does not support `Cas` and `Cad` methods.

Reply:
i get your point now, maybe i should use `(redis.Client).Do` for instead.

Reply:
> Extended SET XX option

There is a `SetXX` method.

> Cas and Cad

Yes. For Kvrocks specific command, use `Do` method. See also `sint` tests..